llvm/lib/Target/AMDGPU/SILoadStoreOptimizer.cpp

```diff
@@ -1701,17 +1732,33 @@ unsigned SILoadStoreOptimizer::getNewOpcode(const CombineInfo &CI,
       return AMDGPU::S_BUFFER_LOAD_DWORDX8_SGPR_IMM;
     }
   case S_LOAD_IMM:
-    switch (Width) {
-    default:
-      return 0;
-    case 2:
-      return AMDGPU::S_LOAD_DWORDX2_IMM;
-    case 3:
-      return AMDGPU::S_LOAD_DWORDX3_IMM;
-    case 4:
-      return AMDGPU::S_LOAD_DWORDX4_IMM;
-    case 8:
-      return AMDGPU::S_LOAD_DWORDX8_IMM;
+    // For targets that support XNACK replay, use the constrained load opcode.
+    if (STI && STI->hasXnackReplay()) {
+      switch (Width) {
```

Comment:
You can check alignment on the first load if MMO is available and avoid producing _ec version if it is sufficient.

Reply:
I guess currently the merged load is always under-aligned. While combining the MMOs, currently the alignment is picked from the first MMO and that'd definitely be smaller than the natural align requirement for the new load. That's one reason I conservatively want to emit _ec equivalent when XNACK is enabled.

Reply:
> currently the alignment is picked from the first MMO and that'd definitely be smaller than the natural align requirement for the new load

You don't know that - the alignment in the first MMO will be whatever alignment the compiler could deduce, which could be large.

Reply:
> > currently the alignment is picked from the first MMO and that'd definitely be smaller than the natural align requirement for the new load
> 
> You don't know that - the alignment in the first MMO will be whatever alignment the compiler could deduce, which could be large, e.g. if the pointer used for the first load was known to have a large alignment.

Are you suggesting to check the alignment in the first MMO and see if it is still the preferred alignment for the merge-load? 
Use the _ec if the alignment is found to be smaller than the expected value.

Reply:
> > currently the alignment is picked from the first MMO and that'd definitely be smaller than the natural align requirement for the new load
> 
> You don't know that - the alignment in the first MMO will be whatever alignment the compiler could deduce, which could be large, e.g. if the pointer used for the first load was known to have a large alignment.

Moreover, it can easily be as large as a page. In a case of scalar load and kernarg.